src/librustc/ty/error.rs

```diff
@@ -247,20 +247,32 @@ impl<'tcx> ty::TyS<'tcx> {
 }
 
 impl<'tcx> TyCtxt<'tcx> {
-    pub fn note_and_explain_type_err(self,
-                                     db: &mut DiagnosticBuilder<'_>,
-                                     err: &TypeError<'tcx>,
-                                     sp: Span) {
+    pub fn note_and_explain_type_err(
+        self,
+        db: &mut DiagnosticBuilder<'_>,
+        err: &TypeError<'tcx>,
+        sp: Span,
+    ) {
         use self::TypeError::*;
 
-        match err.clone() {
+        match err {
             Sorts(values) => {
                 let expected_str = values.expected.sort_string(self);
                 let found_str = values.found.sort_string(self);
                 if expected_str == found_str && expected_str == "closure" {
                     db.note("no two closures, even if identical, have the same type");
                     db.help("consider boxing your closure and/or using it as a trait object");
                 }
+                if expected_str == found_str && expected_str == "opaque type" { // Issue #63167
+                    db.note("distinct uses of `impl Trait` result in different opaque types");
+                    let e_str = values.expected.to_string();
+                    let f_str = values.found.to_string();
+                    if &e_str == &f_str && &e_str == "impl std::future::Future" {
+                        // FIXME: use non-string based check.
+                        db.help("if both futures resolve to the same type, consider `await`ing \
```

Comment:
```suggestion
                        db.help("if both `Future`s have the same `Output` type, consider `.await`ing \
```

Reply:
I'm not entirely happy with the `help`, but it's a relatively good stop gap for now. I'd like the experience in 1.39 to be as good as possible, so I'll make a round for the async/await diagnostics tickets to, if not fix them, have a stop gap like this one.

Reply:
Sure, if you don't think my suggestion is an improvement then let's skip it... but do you? :P

Reply:
The suggestion is good, I'm just not too happy with not actually checking the `Output` nor the mildly confusing "consider...". But it'll do for now.